File: BOJ/yein/8_week/지름길.py
# 1446 지름길

N, D = map(int,input().split())

# 지름길 arr
arr = [list(map(int, input().split())) for _ in range(N)]

# 지름길을 출발 지점 기준으로 정렬
arr.sort()

# 지름길 없이 오는 경우를 뚜벅이라고 함 (차 타고 오지만 아무튼 그럼)
dp = [i for i in range(D+1)] # 전부 뚜벅이로 초기화

# 지름길 arr용 인덱스
k=0
for i in range(D+1):
    dp[i] = min(dp[i-1]+1, dp[i]) # 전 칸 + 1 또는 뚜벅이
    while (k<N): # N은 지름길의 총 개수
        if(i==arr[k][0]): # 지름길이 나 있는 길
            if(arr[k][1]<=D): # 지름길이 목표지점을 벗어난다면 무시
                dp[arr[k][1]] = min(dp[i]+arr[k][2], dp[arr[k][1]]) # 지름길이 뚜벅이보다 먼 경우가 있음. min을 통해 확실히 더 짧은 길 검증.
            k+=1
        else: break

print(dp[D])

File: BOJ/jeongbeen/7_week/지름길.java
import java.io.*;
import java.util.*;

public class 지름길 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] s = br.readLine().split(" ");
        int N = Integer.parseInt(s[0]);
        int D = Integer.parseInt(s[1]);

        ArrayList<ArrayList<Integer>> arr = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            String[] ss = br.readLine().split(" ");
            ArrayList<Integer> list = new ArrayList<>();
            for (int j = 0; j < ss.length; j++) {
                list.add(Integer.parseInt(ss[j]));
            }
            if (list.get(1) <= D && (list.get(1) - list.get(0) - list.get(2) > 0)) arr.add(list);
        }

        int[] dist = new int[D + 1];
        for (int i = 1; i <= D; i++) {
            int min = dist[i - 1] + 1;
            for (ArrayList<Integer> list : arr) {
                if (list.get(1) == i) {
                    min = Math.min(dist[list.get(0)] + list.get(2), min);
                }
            }
            dist[i] = min;
        }
        bw.write(dist[D] + "\n");
        bw.flush();
    }
}
